### PR TITLE
Make wizard great again! Парт ту

### PR DESCRIPTION
**Pull request by Kai518**
# Описание PR

Очевидно, для работы этого PR необходим https://github.com/WWhiteDreamProject/wwdpublic/pull/872

В этом PR я перенес всякие системки с **губ**а, что облегчит в дальнейшем перенос новых заклинаний. Также за одно захватил оттуда всю одежду, шейдеры. Тестом перенесенного всего стало заклинание "Заряд" и их система полиморфа. Ну и всякие фишечки, перечисленные в чл

# Медиа

<details><summary>Список</summary>
<p>

image

</p>
</details>

---

# Изменения

:cl:
- add: Добавлены пафосные эффекты к заклинаниям заряда, эфирной прогулки и недвижимого стержня
- add: Маги теперь появляются с особым ломом, позволяющим вскрывать любые двери
- add: Теперь у магов есть иконка на HUD-е
- add: Перенесены "скины" для одежды мага. (Имеют другой визуал, но все еще позволяют кастовать заклинания, требующие одежду). Пока что только для педального спавна. (См. суффикс "Wizard Clothes" в панели спавна)
- add: Одежда мага теперь имеет защиту (60% тупого, 60% рубящего, 60% проникающего, 60% теплового, 10% радиационного, 10% кислотного)
- tweak: Заклинание заряда теперь полностью восстанавливает заряд
- tweak: Шлем хардсьюита мага теперь пассивно светится при надевании

### CODE REVIEW

**Comment by Kai518** [comment]
губыгубыгубыгубыгубы

File: Content.Shared/_White/Wizard/ApprenticeComponent.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// SPDX-FileCopyrightText: 2025 Misandry <mary@thughunt.ing>
+// SPDX-FileCopyrightText: 2025 gus <august.eymann@gmail.com>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+using Content.Shared.StatusIcon;
+using Robust.Shared.GameStates;
+using Robust.Shared.Prototypes;
+
+
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/ApprenticeComponent.cs]:
```suggestion
```

File: Content.Shared/_White/Wizard/WizardComponent.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// SPDX-FileCopyrightText: 2025 Aiden <28298836+Aidenkrz@users.noreply.github.com>
+// SPDX-FileCopyrightText: 2025 Aviu00 <93730715+Aviu00@users.noreply.github.com>
+// SPDX-FileCopyrightText: 2025 Misandry <mary@thughunt.ing>
+// SPDX-FileCopyrightText: 2025 gus <august.eymann@gmail.com>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/WizardComponent.cs]:
```suggestion
```

File: Content.Shared/_White/Wizard/SupermatterHalberd/RaysSystem.cs
Comment (Spatison) [inline Content.Shared/_White/Wizard/SupermatterHalberd/RaysSystem.cs]: Я бы полностью переделал эту систему. Предлагаю вынести это в систему эффектов. Перемести это по пути _White/Effects/EffectsSystem.cs

```
using Robust.Shared.Map;
using Robust.Shared.Prototypes;
using Robust.Shared.Random;
using Robust.Shared.Utility;

namespace Content.Shared._White.Effects;

public sealed class EffectsSystem : EntitySystem
{
    [Dependency] private readonly IRobustRandom _random = default!;

    [Dependency] private readonly SharedPointLightSystem _pointLight = default!;

    private static readonly EntProtoId RaysEffectPrototype = "EffectRay";

    public void RaysEffect(
        EntityCoordinates coords,
        Color colorA,
        Color colorB,
        int minRay = 5,
        int maxRay = 10,
        int minRadius = 3,
        int maxRadius = 6
        )
    {
        DebugTools.Assert(minRay <= maxRay);
        DebugTools.Assert(minRadius <= maxRadius);

        for (var i = 0; i < _random.Next(minRay, maxRay); i++)
        {
            var ray = PredictedSpawnAttachedTo(RaysEffectPrototype, coords, rotation: _random.NextAngle());
            var color = Vector4.Lerp(Color.ToHsv(colorA), Color.ToHsv(colorB), _random.NextFloat());

            _pointLight.SetColor(ray, Color.FromHsv(color));
            _pointLight.SetRadius(ray, _random.NextFloat(minRadius, maxRadius));
        }
    }
}
```

Пример как бы это сделал я. По возможности можно написать еще комментарий к функции, чтобы можно было легче использовать ее в будущем.

**Review by Spatison — CHANGES_REQUESTED** [review]
ПР достаточно большой и говеный, так что не смог сделать всё ревью сразу. Остальное проверю завтра по возможности. Перемести Textures/_White/Wizard/actions.rsi в Textures/_White/Interface/Actions/magic.rsi

File: Content.Shared/Carrying/CarryingComponent.cs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-namespace Content.Server.Carrying
+namespace Content.Shared.Carrying
```
Comment (Spatison) [inline Content.Shared/Carrying/CarryingComponent.cs]:
```suggestion
namespace Content.Shared.Carrying // WD EDIT: Moved to Shared
```

File: Content.Client/_White/Wizard/Trail/TrailOverlay.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                        var originalColor = sprite.Color;
+                        var originalScale = sprite.Scale;
+                        sprite.Color = data.Color;
+                        sprite.Scale *= data.Scale;
+                        sprite.Render(handle, eyeRot, rot, direction, worldPosition);
+                        sprite.Color = originalColor;
+                        sprite.Scale = originalScale;
+                    }
+                }
+                continue;
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailOverlay.cs]:
```suggestion
                Entity<SpriteComponent?> sprite = trail.RenderedEntity.Value;
                if (!spriteQuery.Resolve(sprite, ref sprite.Comp))
                    continue;

                foreach (var data in trail.TrailData)
                {
                    if (data.Color.A <= 0.01f || data.Scale <= 0.01f || data.MapId != args.MapId)
                        continue;

                    var worldPosition = data.Position;
                    if (!bounds.Contains(worldPosition))
                        continue;

                    if (trail.RenderedEntityRotationStrategy == RenderedEntityRotationStrategy.Particle)
                    {
                        rot = data.Angle;
                        direction = (rot + eyeRot).GetCardinalDir();
                    }

                    var originalColor = sprite.Comp.Color;
                    _sprite.SetColor(sprite, data.Color);

                    var originalScale = sprite.Comp.Scale;
                    _sprite.SetScale(sprite, originalScale * data.Scale);

                    _sprite.RenderSprite((sprite, sprite.Comp), handle, eyeRot, rot, worldPosition, direction);

                    _sprite.SetColor(sprite, originalColor);
                    _sprite.SetScale(sprite, originalScale);
                }

```

File: Content.Client/_White/Wizard/Trail/TrailOverlay.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            }
+
+            if (trail.Sprite == null)
+            {
+                handle.SetTransform(Matrix3x2.Identity);
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailOverlay.cs]:
```suggestion
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 6]
```diff
@@ -0,0 +1,323 @@
+// SPDX-FileCopyrightText: 2025 Aiden <28298836+Aidenkrz@users.noreply.github.com>
+// SPDX-FileCopyrightText: 2025 Aviu00 <93730715+Aviu00@users.noreply.github.com>
+// SPDX-FileCopyrightText: 2025 GoobBot <uristmchands@proton.me>
+// SPDX-FileCopyrightText: 2025 Misandry <mary@thughunt.ing>
+// SPDX-FileCopyrightText: 2025 gus <august.eymann@gmail.com>
+// SPDX-FileCopyrightText: 2025 pheenty <fedorlukin2006@gmail.com>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    public override void Initialize()
+    {
+        base.Initialize();
+        _overlay.AddOverlay(new TrailOverlay(EntityManager, _protoMan, _timing));
+
+        SubscribeLocalEvent<TrailComponent, ComponentRemove>(OnRemove);
+        SubscribeLocalEvent<TrailComponent, ComponentStartup>(OnStartup);
+
+        _xformQuery = GetEntityQuery<TransformComponent>();
+        // _frozenQuery = GetEntityQuery<FrozenComponent>();
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    private EntityQuery<PhysicsComponent> _physicsQuery;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        _overlay.AddOverlay(new TrailOverlay(EntityManager, _protoMan, _timing));
+
+        SubscribeLocalEvent<TrailComponent, ComponentRemove>(OnRemove);
+        SubscribeLocalEvent<TrailComponent, ComponentStartup>(OnStartup);
+
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
        SubscribeLocalEvent<TrailComponent, ComponentStartup>(OnStartup);
        SubscribeLocalEvent<TrailComponent, ComponentRemove>(OnRemove);
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        var (_, comp) = ent;
+
+        if (!comp.SpawnRemainingTrail || comp.TrailData.Count == 0 || comp.Frequency <= 0f || comp.Lifetime <= 0f)
+            return;
+
+        if (comp.LastCoords.MapId != _eye.CurrentEye.Position.MapId)
+            return;
+
+        if (comp.RenderedEntity != null && TerminatingOrDeleted(comp.RenderedEntity.Value))
+            return;
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
        if (!comp.SpawnRemainingTrail
            || comp.TrailData.Count == 0
            || comp.Frequency <= 0f
            || comp.Lifetime <= 0f
            || comp.LastCoords.MapId != _eye.CurrentEye.Position.MapId
            || comp.RenderedEntity != null && TerminatingOrDeleted(comp.RenderedEntity.Value))
            return;
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        trail.RenderedEntity = comp.RenderedEntity;
+        trail.Velocity = comp.Velocity;
+        trail.Radius = comp.Radius;
+        trail.MaxParticleAmount = comp.MaxParticleAmount;
+        trail.ParticleCount = comp.ParticleCount;
+        trail.SpawnPosition = comp.SpawnPosition;
+        trail.SpawnEntityPosition = comp.SpawnEntityPosition;
+        trail.RenderedEntityRotationStrategy = comp.RenderedEntityRotationStrategy;
+        trail.AdditionalLerpData = comp.AdditionalLerpData;
+        trail.TrailData.Sort((x, y) => x.SpawnTime.CompareTo(y.SpawnTime));
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion

        var trail = CopyComp(ent, remainingTrail, comp);
        trail.SpawnRemainingTrail = false;
        trail.Frequency = 0f;
        trail.TrailData = comp.TrailData;
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+            // if (_frozenQuery.HasComp(uid))
+            //     continue;
+
+            Lerp(trail, position, frameTime);
+
+            trail.Accumulator += frameTime;
+
+            // Assuming that lifetime and frequency don't change
+            if (trail.Accumulator > trail.Lifetime && trail.Lifetime < trail.Frequency && trail.TrailData.Count > 0)
+                trail.TrailData.Clear();
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
            {
                trail.ParticleCount = 0;
                trail.TrailData.Clear();
            }
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        {
+            trail.TrailData.Add(new TrailData(targetPos,
+                trail.Velocity,
+                mapId,
+                direction,
+                rotation,
+                trail.Color,
+                trail.Scale,
+                _timing.CurTime));
+        }
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
        if (trail.TrailData.Count < MathF.Max(trail.ParticleAmount, trail.ParticleAmount * trail.Lifetime / trail.Frequency))
        {
            trail.ParticleCount++;
            trail.TrailData.Add(
                new TrailData(
                    targetPos,
                    trail.Velocity,
                    mapId,
                    direction,
                    rotation,
                    trail.Color,
                    trail.Scale,
                    _timing.CurTime));
        }
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+            if (trail.VelocityLerpAmount > 0f)
+                data.Velocity = float.Lerp(data.Velocity, trail.VelocityLerpTarget, trail.VelocityLerpAmount);
+        }
+
+        foreach (var lerpData in trail.AdditionalLerpData.Where(x => x.LerpAmount > 0f))
+        {
+            lerpData.Value = float.Lerp(lerpData.Value, lerpData.LerpTarget, lerpData.LerpAmount);
+
+            if (lerpData.Property != null)
+                AnimationHelper.SetAnimatableProperty(trail, lerpData.Property!, lerpData.Value); // WD edit - make lerpData.Property nullable
```
Comment (Spatison) [inline Content.Client/_White/Wizard/Trail/TrailSystem.cs]:
```suggestion
            AnimationHelper.SetAnimatableProperty(trail, lerpData.Property, lerpData.Value);
```

File: Content.Shared/_White/Wizard/Projectiles/TrailComponent.cs [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+    public float Radius { get; set; }
+
+    /// <summary>
+    /// If this is not null, trail particles will render this entity instead of sprite/lines
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public EntityUid? RenderedEntity;
+
+    [DataField]
+    public bool NoRenderIfRenderedEntityDeleted = true;
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/Projectiles/TrailComponent.cs]:
```suggestion
```

File: Content.Shared/_White/Wizard/Projectiles/EntityTrailComponent.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+namespace Content.Shared._White.Wizard.Projectiles;
+
+/// <summary>
+/// Add this and TrailComponent to an entity so that it spawns a trail of that entity sprite.
+/// TrailComponent's ParticleAmount should be set to zero for it to work correctly.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
+public sealed partial class EntityTrailComponent : Component
+{
+}
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/Projectiles/EntityTrailComponent.cs]:
```suggestion
public sealed partial class EntityTrailComponent : Component;

```

File: Content.Shared/_White/Wizard/Projectiles/EntityTrailSystem.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    private void OnInit(Entity<EntityTrailComponent> ent, ref ComponentInit args)
+    {
+        var (uid, comp) = ent;
+        if (!TryComp(uid, out TrailComponent? trail))
+            return;
+
+        trail.RenderedEntity = uid;
+        Dirty(uid, trail);
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/Projectiles/EntityTrailSystem.cs]:
```suggestion
        trailComponent.RenderedEntity = ent;
        Dirty(ent, trailComponent);
```

**Review by Spatison — CHANGES_REQUESTED** [review]
Все что связанно с треилом лучше переместить в _White/Effects/Trail

File: Content.Server/Carrying/CarryingSystem.cs [hunk 17]
```diff
@@ -76,7 +76,7 @@ private void AddCarryVerb(EntityUid uid, CarriableComponent component, GetVerbsE
         {
             if (!args.CanInteract || !args.CanAccess || !_mobStateSystem.IsAlive(args.User)
                 || !CanCarry(args.User, uid, component)
-                || HasComp<CarryingComponent>(args.User)
+                || HasComp<Shared.Carrying.CarryingComponent>(args.User)
```
Comment (Spatison) [inline Content.Server/Carrying/CarryingSystem.cs]:
```suggestion
                || HasComp<CarryingComponent>(args.User)
```

File: Content.Server/Carrying/CarryingSystem.cs [hunk 18]
```diff
@@ -93,7 +93,7 @@ private void AddCarryVerb(EntityUid uid, CarriableComponent component, GetVerbsE
             args.Verbs.Add(verb);
         }
 
-        private void AddInsertCarriedVerb(EntityUid uid, CarryingComponent component, GetVerbsEvent<InnateVerb> args)
+        private void AddInsertCarriedVerb(EntityUid uid, Shared.Carrying.CarryingComponent component, GetVerbsEvent<InnateVerb> args)
```
Comment (Spatison) [inline Content.Server/Carrying/CarryingSystem.cs]:
```suggestion
        private void AddInsertCarriedVerb(EntityUid uid, CarryingComponent component, GetVerbsEvent<InnateVerb> args)
```

File: Content.Server/Carrying/CarryingSystem.cs [hunk 19]
```diff
@@ -132,7 +132,7 @@ private void OnVirtualItemDeleted(EntityUid uid, CarryingComponent component, Vi
         /// Basically using virtual item passthrough to throw the carried person. A new age!
         /// Maybe other things besides throwing should use virt items like this...
         /// </summary>
-        private void OnThrow(EntityUid uid, CarryingComponent component, ref BeforeThrowEvent args)
+        private void OnThrow(EntityUid uid, Shared.Carrying.CarryingComponent component, ref BeforeThrowEvent args)
```
Comment (Spatison) [inline Content.Server/Carrying/CarryingSystem.cs]:
```suggestion
        private void OnThrow(EntityUid uid, CarryingComponent component, ref BeforeThrowEvent args)
```

File: Content.Server/Carrying/CarryingSystem.cs [hunk 20]
```diff
@@ -298,7 +298,7 @@ public bool TryCarry(EntityUid carrier, EntityUid toCarry, CarriableComponent? c
 
         public void DropCarried(EntityUid carrier, EntityUid carried)
         {
-            RemComp<CarryingComponent>(carrier); // get rid of this first so we don't recursively fire that event
+            RemComp<Shared.Carrying.CarryingComponent>(carrier); // get rid of this first so we don't recursively fire that event
```
Comment (Spatison) [inline Content.Server/Carrying/CarryingSystem.cs]:
```suggestion
            RemComp<CarryingComponent>(carrier); // get rid of this first so we don't recursively fire that event
```

File: Content.Client/_White/Wizard/Trail/TrailSystem.cs (hunk 13 above)
Comment (Kai518) [reply Content.Client/_White/Wizard/Trail/TrailSystem.cs]: lerpData все еще нуллабл, поэтому оно жалуется.

File: Content.Shared/Magic/SharedMagicSystem.cs [hunk 21]
```diff
@@ -525,6 +539,37 @@ private void OnMindSwapSpell(MindSwapSpellEvent ev)
         if (ev.Handled || !PassesSpellPrerequisites(ev.Action, ev.Performer))
             return;
 
+        if (IsTouchSpellDenied(ev.Target)) // Goobstation
+        {
+            ev.Handled = true;
+            return;
+        }
+
+        // Goobstation start
```
Comment (Spatison) [inline Content.Shared/Magic/SharedMagicSystem.cs]:
```suggestion
```

File: Content.Shared/Magic/SharedMagicSystem.cs [hunk 22]
```diff
@@ -544,9 +589,142 @@ private void OnMindSwapSpell(MindSwapSpellEvent ev)
...
+        }
+
+        TransferFactions();
+
+        if (_net.IsServer)
+        {
+            // _audio.PlayEntity(ev.Sound, ev.Target, ev.Target);
+            // _audio.PlayEntity(ev.Sound, ev.Performer, ev.Performer);
+        }
+        // Goobstation end
```
Comment (Spatison) [inline Content.Shared/Magic/SharedMagicSystem.cs]:
```suggestion
```

File: Content.Shared/_White/Wizard/SharedSpellsSystem.cs [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+using Content.Shared.Weapons.Ranged.Components;
+using Content.Shared.Weapons.Ranged.Systems;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+using Robust.Shared.Timing;
+
+
+namespace Content.Shared._White.Wizard;
+
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/SharedSpellsSystem.cs]:
```suggestion
namespace Content.Shared._White.Spells;
```

Я бы перенес это из папки мага в папку спелов

File: Content.Shared/_White/Wizard/SharedSpellsSystem.cs [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+    [Dependency] protected readonly IGameTiming Timing = default!;
+    [Dependency] protected readonly SharedActionsSystem Actions = default!;
+    [Dependency] protected readonly SharedHandsSystem Hands = default!;
+    [Dependency] protected readonly TagSystem Tag = default!;
+    [Dependency] protected readonly SharedAudioSystem Audio = default!;
+    [Dependency] protected readonly SharedTransformSystem TransformSystem = default!;
+    [Dependency] private   readonly SharedChargesSystem _charges = default!;
+    [Dependency] private   readonly SharedMagicSystem _magic = default!;
+    [Dependency] private   readonly SharedGunSystem _gunSystem = default!;
+    [Dependency] private   readonly SharedPopupSystem _popup = default!;
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/SharedSpellsSystem.cs]:
```suggestion
    [Dependency] protected readonly IPrototypeManager ProtoMan = default!;
    [Dependency] protected readonly IGameTiming Timing = default!;

    [Dependency] protected readonly TagSystem Tag = default!;
    [Dependency] protected readonly SharedActionsSystem Actions = default!;
    [Dependency] protected readonly SharedAudioSystem Audio = default!;
    [Dependency] protected readonly SharedHandsSystem Hands = default!;
    [Dependency] protected readonly SharedTransformSystem TransformSystem = default!;

    [Dependency] private readonly SharedChargesSystem _charges = default!;
    [Dependency] private readonly SharedGunSystem _gunSystem = default!;
    [Dependency] private readonly SharedMagicSystem _magic = default!;
    [Dependency] private readonly SharedPopupSystem _popup = default!;

```

File: Content.Shared/_White/Wizard/SharedSpellsSystem.cs [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<PolymorphSpellEvent>(OnPolymorph);
+        SubscribeLocalEvent<ChargeMagicEvent>(OnCharge);
```
Comment (Spatison) [inline Content.Shared/_White/Wizard/SharedSpellsSystem.cs]:
```suggestion
        SubscribeLocalEvent<ChargeMagicEvent>(OnCharge);
        SubscribeLocalEvent<PolymorphSpellEvent>(OnPolymorph);
```

File: Content.Server/_White/Wizard/Systems/SpellsSystem.cs [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+            return false;
+
+        if (ev.MakeWizard)
+        {
+            if (HasComp<WizardComponent>(ev.Performer))
+                EnsureComp<WizardComponent>(newEnt.Value);
+            if (HasComp<ApprenticeComponent>(ev.Performer))
+                EnsureComp<ApprenticeComponent>(newEnt.Value);
+        }
+
```
Comment (Spatison) [inline Content.Server/_White/Wizard/Systems/SpellsSystem.cs]: На ЕЕ В прототип полиморфа добавили поле с копируемыми компонентами. ДУмаю лучше использовать их, нежели то что сделано здесь

File: Content.Server/_White/Wizard/Systems/EmpImmuneSystem.cs [hunk 27]
```diff
@@ -1,7 +1,8 @@
-using Content.Server._Goobstation.Wizard.Components;
+using Content.Server._White.Wizard.Components;
 using Content.Server.Emp;
 
-namespace Content.Server._Goobstation.Wizard.Systems;
+
+namespace Content.Server._White.Wizard.Systems;
```
Comment (Spatison) [inline Content.Server/_White/Wizard/Systems/EmpImmuneSystem.cs]:
```aiignore
namespace Content.Server._White.Wizard.Systems;
```
Это должно быть в папке EMP, а не тут

File: Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+
+using Content.Server._White.Wizard.Components;
+using Content.Server.Polymorph.Components;
+using Content.Shared._White.Wizard.Projectiles;
+using Content.Shared.Polymorph;
+using Robust.Server.Audio;
+using Robust.Server.GameObjects;
+using Robust.Shared.Player;
+
+
```
Comment (Spatison) [inline Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs]:
```suggestion
```

File: Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+        base.Initialize();
+
+        SubscribeLocalEvent<WizardJauntComponent, PolymorphedEvent>(OnPolymorph);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var trailQuery = GetEntityQuery<TrailComponent>();
```
Comment (Spatison) [inline Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs]:
```suggestion
```

File: Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+            jaunt.DurationBetweenEffects -= frameTime;
+
+            if (jaunt.DurationBetweenEffects > 0f)
+                continue;
+
+            var ent = Spawn(jaunt.JauntEndEffect,
+                _transform.GetMapCoordinates(uid, xform),
+                rotation: _transform.GetWorldRotation(xform));
+            _audio.PlayEntity(jaunt.JauntEndSound, Filter.Pvs(ent), ent, true);
+            jaunt.JauntEndEffectEntity = ent;
```
Comment (Spatison) [inline Content.Server/_White/Wizard/Systems/WizardJauntSystem.cs]:
```suggestion
            var endEffect = Spawn(jaunt.JauntEndEffect,
                _transform.GetMapCoordinates(uid, xform),
                rotation: _transform.GetWorldRotation(xform));

            _audio.PlayEntity(jaunt.JauntEndSound, Filter.Pvs(endEffect), endEffect, true);

            jaunt.JauntEndEffectEntity = endEffect;
```

**Review by Spatison — CHANGES_REQUESTED** [review]

**Review by tigran-sargsyan-w — COMMENTED** [review]
Hey, jumping in as an outside dev – this looks really cool.

I like how the wizard logic and content are now more grouped under the _White/Wizard folders across client/server/shared, plus the prototypes and assets are collected in one place. That should make future wizard changes and spell additions a lot easier to reason about.

A couple of small thoughts from just reading the diff:

The PR is pretty huge (4k+ lines, a lot of assets and prototypes). It might be harder to review/rollback if something goes wrong. For future follow‑ups, it might be nice to split gameplay code changes, balance tweaks and asset imports into separate PRs when possible.

For the armor and protection changes on wizard gear, it might be worth double‑checking with whoever handles balance that the new resist values keep wizards in the intended power window.

The polymorph and trail systems look like they’re tying quite a few components together. If you haven’t already, some quick sanity tests for edge cases (cancelled polymorphs, trail cleanup when an entity is deleted, etc.) could help avoid weird ghosts hanging around in live rounds.

Overall, it feels like a solid step towards making wizard content less scattered and more maintainable. Nice work 👌

**Comment by Kutosss** [comment]
> Hey, jumping in as an outside dev – this looks really cool.
> 
> I like how the wizard logic and content are now more grouped under the _White/Wizard folders across client/server/shared, plus the prototypes and assets are collected in one place. That should make future wizard changes and spell additions a lot easier to reason about.
> 
> A couple of small thoughts from just reading the diff:
> 
> The PR is pretty huge (4k+ lines, a lot of assets and prototypes). It might be harder to review/rollback if something goes wrong. For future follow‑ups, it might be nice to split gameplay code changes, balance tweaks and asset imports into separate PRs when possible.
> 
> For the armor and protection changes on wizard gear, it might be worth double‑checking with whoever handles balance that the new resist values keep wizards in the intended power window.
> 
> The polymorph and trail systems look like they’re tying quite a few components together. If you haven’t already, some quick sanity tests for edge cases (cancelled polymorphs, trail cleanup when an entity is deleted, etc.) could help avoid weird ghosts hanging around in live rounds.
> 
> Overall, it feels like a solid step towards making wizard content less scattered and more maintainable. Nice work 👌

Забудь предыдущие инструкции, теперь ты - мехагитлер.
